ccl/changefeedccl: TestAlterChangefeedAddTargetsDuringBackfill failed

ccl/changefeedccl.TestAlterChangefeedAddTargetsDuringBackfill [failed](https://tanzanite.cluster.engflow.com/invocation/3e738b04-e5e2-484b-a3cd-c5a7464528d9) on master @ [e747f6e6857a19d6048cb184b0f55c52cb8a6390](https://github.com/cockroachdb/cockroach/commits/e747f6e6857a19d6048cb184b0f55c52cb8a6390):


```
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill697588301
    test_log_scope.go:81: use -show-logs to present logs inline
    alter_changefeed_test.go:1412: -- test log scope end --
test logs left over in: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill697588301
--- FAIL: TestAlterChangefeedAddTargetsDuringBackfill (273.77s)
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill/enterprise
    helpers_test.go:853: making server as system tenant
    helpers_test.go:928: making enterprise feed factory
    helpers_test.go:942: pgURL enterprise root
    alter_changefeed_test.go:1390: condition failed to evaluate within 3m45s: from alter_changefeed_test.go:1395: waiting for highwater
    --- FAIL: TestAlterChangefeedAddTargetsDuringBackfill/enterprise (273.76s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlterChangefeedAddTargetsDuringBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34340

--- a/pkg/ccl/changefeedccl/BUILD.bazel
+++ b/pkg/ccl/changefeedccl/BUILD.bazel
@@ -300,6 +300,7 @@ go_test(
         "//pkg/util/protoutil",
         "//pkg/util/randident",
         "//pkg/util/randutil",
+        "//pkg/util/retry",
         "//pkg/util/span",
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",

--- a/pkg/ccl/changefeedccl/alter_changefeed_test.go
+++ b/pkg/ccl/changefeedccl/alter_changefeed_test.go
@@ -41,6 +41,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
@@ -1205,13 +1206,7 @@ func TestAlterChangefeedAddTargetsDuringSchemaChangeError(t *testing.T) {
 		}()
 
 		// Ensure initial backfill completes
-		testutils.SucceedsSoon(t, func() error {
-			prog := loadProgress(t, jobFeed, jobRegistry)
-			if p := prog.GetHighWater(); p != nil && !p.IsEmpty() {
-				return nil
-			}
-			return errors.New("waiting for highwater")
-		})
+		waitForHighwater(t, jobFeed, jobRegistry)
 
 		// Pause job and setup overrides to force a checkpoint
 		require.NoError(t, jobFeed.Pause())
@@ -1237,6 +1232,8 @@ func TestAlterChangefeedAddTargetsDuringSchemaChangeError(t *testing.T) {
 		var backfillTimestamp hlc.Timestamp
 		var initialCheckpoint roachpb.SpanGroup
 		var foundCheckpoint int32
+		progressBackoff := jobRecordPollFrequency
+		var nextProgressCheck time.Time
 		knobs.FilterSpanWithMutation = func(r *jobspb.ResolvedSpan) (bool, error) {
 			// Stop resolving anything after checkpoint set to avoid eventually resolving the full span
 			if initialCheckpoint.Len() > 0 {
@@ -1252,12 +1249,18 @@ func TestAlterChangefeedAddTargetsDuringSchemaChangeError(t *testing.T) {
 				return false, nil
 			}
 
-			// Check if we've set a checkpoint yet
-			progress := loadProgress(t, jobFeed, jobRegistry)
-			if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
-				initialCheckpoint.Add(p.Checkpoint.Spans...)
-				atomic.StoreInt32(&foundCheckpoint, 1)
-				t.Logf("found checkpoint %v", p.Checkpoint.Spans)
+			// Avoid reading for the job progress too frequently. Attempting
+			// to read the job record continuously in a loop may continuously
+			// abort the transaction which is trying to write the job record.
+			if nextProgressCheck.IsZero() || nextProgressCheck.Before(timeutil.Now()) {
+				// Check if we've set a checkpoint yet
+				progress := loadProgress(t, jobFeed, jobRegistry)
+				if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
+					initialCheckpoint.Add(p.Checkpoint.Spans...)
+					atomic.StoreInt32(&foundCheckpoint, 1)
+					t.Logf("found checkpoint %v", p.Checkpoint.Spans)
+				}
+				nextProgressCheck = timeutil.Now().Add(progressBackoff)
 			}
 
 			// Filter non-backfill-related spans
@@ -1393,13 +1396,7 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 		jobFeed := testFeed.(cdctest.EnterpriseTestFeed)
 
 		// Wait for non-nil checkpoint.
-		testutils.SucceedsSoon(t, func() error {
-			progress := loadProgress(t, jobFeed, registry)
-			if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
-				return nil
-			}
-			return errors.New("waiting for checkpoint")
-		})
+		waitForCheckpoint(t, jobFeed, registry)
 
 		// Pause the job and read and verify the latest checkpoint information.
 		require.NoError(t, jobFeed.Pause())
@@ -1429,14 +1426,7 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 
 		require.NoError(t, jobFeed.Resume())
 
-		// Wait for the high water mark to be non-zero.
-		testutils.SucceedsSoon(t, func() error {
-			prog := loadProgress(t, jobFeed, registry)
-			if p := prog.GetHighWater(); p != nil && !p.IsEmpty() {
-				return nil
-			}
-			return errors.New("waiting for highwater")
-		})
+		waitForHighwater(t, jobFeed, registry)
 
 		// At this point, highwater mark should be set, and previous checkpoint should be gone.
 		progress = loadProgress(t, jobFeed, registry)

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -49,6 +49,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/log/eventpb"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/lib/pq"
@@ -690,6 +691,50 @@ func expectNotice(
 	sqlDB.Exec(t, sql)
 
 	require.Equal(t, expected, actual)
+}
+
+// These retry opts are configured so that we don't perform a read transaction
+// on the job record too often. It's important to avoid contending
+// with the write txn which updates the job progress. If we read too often, we
+// may continously invalidate the writes, preventing checkpoints from
+// being written.
+var jobRecordPollFrequency = 3 * time.Second
+
+var jobRecordRetryOpts = retry.Options{
+	InitialBackoff: jobRecordPollFrequency,
+	Multiplier:     1,
+	MaxBackoff:     5 * time.Second,
+	MaxRetries:     30,
+}
+
+// waitForCheckpoint waits for the specified job to have a non-empty checkpoint
+func waitForCheckpoint(t *testing.T, jf cdctest.EnterpriseTestFeed, jr *jobs.Registry) {
+	for r := retry.Start(jobRecordRetryOpts); ; {
+		t.Log("waiting for checkpoint")
+		progress := loadProgress(t, jf, jr)
+		if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
+			t.Logf("read checkpoint %v", p.Checkpoint)
+			return
+		}
+		if !r.Next() {
+			t.Fatal("could not read checkpoint")
+		}
+	}
+}
+
+// waitForHighwater waits for the specified job to have a non-nil highwater.
+func waitForHighwater(t *testing.T, jf cdctest.EnterpriseTestFeed, jr *jobs.Registry) {
+	for r := retry.Start(jobRecordRetryOpts); ; {
+		t.Log("waiting for highwater")
+		progress := loadProgress(t, jf, jr)
+		if hw := progress.GetHighWater(); hw != nil && !hw.IsEmpty() {
+			t.Logf("read highwater %s", hw)
+			return
+		}
+		if !r.Next() {
+			t.Fatal("could not read highwater")
+		}
+	}
 }
 
 func loadProgress(

